### DISCUSSION TITLE
Learning how to use

### DISCUSSION THREAD

**Discussion by josuedapaco** [pyscada/PyScada#150] [Q&A]
I'm using PyScada for my thesis, but I want to learn to use it well. Is it possible that you can help me with that, please?

**Comment by josuedapaco** [comment]
I need to know well how to create HMIs and how I can add graphs.

**Reply by clavay** [reply]
First you need to create devices (see [here](https://pyscada.readthedocs.io/en/main/backend.html#add-a-new-device)) and variables (see [here](https://pyscada.readthedocs.io/en/main/backend.html#add-a-new-variable)).

Then to create the HMI, it is shortly explained [here](https://pyscada.readthedocs.io/en/main/backend.html#short-instructions-to-build-the-user-hmi-frontend)  :
1. create a page,
2. create a view and add the page you created to this view,
3. create a chart and in the `Y-axis` config select the variable you want to display in `chosen variables`
4. create a widget to put your chart in a page,
5. go to the user interface (refresh it if it is already open).

**Comment by josuedapaco** [comment]
I've been trying to figure out how to add a button, but I'm not able to do it. Please help me with this.

**Reply by clavay** [reply]
If you want to have a button to change a variable state, you need to : 

1. set the variable as `writeable` and set the `value class` to `BOOLEAN`,
2. create a control item in the HMI section, select the previous variable and set it as `control element`,
3. create a control panel and add the previous control item,
4. add a widget to put your control panel in a page.

When you will click on the button in the user interface, you can see in the admin interface a new item in `device write task` and see if PyScada succeed to write the new variable value to the related device (`done` checked) or if it failed (`failed` checked).

**Comment by josuedapaco** [comment]
First of all, thank you very much for your help. Is it possible to load images that have one color when the state is zero and another color when the state is one?

**Reply by clavay** [reply]
What do you mean by loading `images that have one color` ?

**Reply by clavay** [reply]
It is possible to change the text and the color of a control item depending on the variable value.

**Reply by josuedapaco** [reply]
For example, the ISA 101 standard specifies that when it is on, it should be white, and when it is off, it should appear dark gray. I already have the images in those colors; I just need to make the white image visible when it is activated and the dark gray one visible when it is deactivated.

**Reply by clavay** [reply]
In PyScada you can create a control item with the `type` as `display value` and create a `display value options` to:

- color = white (`#ffffff`)
- `color only` = True (no text or value will be displayed, only the color)
- click on `Add another Display value color option`, then:
  - `color level` = 1
  - color = gray (`#698194`)

**Reply by josuedapaco** [reply]
I understand, but in that case, it would be a rectangular illustration. What I want to try to do is incorporate images to have an HMI like the one in the following image:

![image](https://github.com/pyscada/PyScada/assets/52479069/32a40616-e116-449c-8a12-a3382cce75aa)
 

What this means is that if I want to illustrate a pump, I will have the graphical representation of the pump, and similarly with a solenoid valve, and so on. I have these images available; I just need to add them to a library and have PyScada access this library.

**Reply by clavay** [reply]
You can achieve something like that using the process flow diagram in the HMI section of PyScada admin.

You will not be able to put some elements (like charts, vertical gauge) on the process flow diagram, but we are open to pull request if you want to implement it.

I don't have time to add these missing elements for the moment.

**Reply by josuedapaco** [reply]
Can you guide me on how to add these images that I mentioned, please? This is what I would like to do for my thesis topic, or how to make a process flow diagram.

**Reply by clavay** [reply]
- Go to the PyScada admin interface.
- In the HMI section add a new process flow diagram.
- set a title, choose a background image, select html or svg (test and see the difference: HTML is not responsive and can display control element, SVG is responsive and cannot display control element)
- Add process flow diagram items to your process flow diagram: a process flow diagram item is a control item with a position relative to the background image.

**Comment by josuedapaco** [comment]
Excuse me, another question, How can I view the code of PyScada, specifically the Django part, to see how it is structured?

**Reply by clavay** [reply]
Here is the code https://github.com/pyscada/PyScada

**Comment by clavay** [comment]
Please close the discussion if you don't have any other question.

**Comment by josuedapaco** [comment]
Sorry, I had to pause the project. I’m trying to connect via Modbus TCP with a LOGO PLC, but it won’t let me read or write to address 8257 to activate a memory. Could you please help me?

**Comment by clavay** [comment]
Is you issue related to pyscada ?
Please give me more details.

**Reply by josuedapaco** [reply]
In “Devices,” I already added the Modbus TCP connection, setting it to read with configuration 0-1-2-3 and the IP address 192.168.100, which is where the PLC is located. Then, I added a write variable as a boolean at address 8257. However, when I press the button, it doesn’t activate that variable.

**Comment by josuedapaco** [comment]
I wanna ask another thing, I have already connected via modbus, but when I make a change it stays charging and never stop, do you know why?![image](https://github.com/user-attachments/assets/b2318e3b-0339-4ff0-b503-9375b7fdaea2) In the picture you can see at the left of Control_Item_2

**Reply by clavay** [reply]
can you show me the logs when you press the button ?

You can use this command:
`tail -fn1000 /var/log/pyscada/pyscada_debug.log`

**Reply by clavay** [reply]
Can you show me what you have in the admin interface in the Device Write Task section ?
